### ISSUE TITLE
[1.13] graphical issue with mounted and drake moving animations (GNA #24919)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1565]
**[Original submission](https://gna.org/bugs/index.php?24919)** by inky on 2016-08-05

Mounted units: Moving them will sometimes cause their sprite to skip around.

Drakes: Moving them will sometimes cause parts of their sprite to be left behind in the hexes they've moved through (these parts disappear once you move the mouse).

(Reproduced on windows 10)
Release: 1.13.5
Priority: 5 - Normal
Severity: 3 - Normal

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-08-05**

**wedge009 wrote:**

> How often does this happen? I wasn't able to get this to occur on my system.

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-08-05**

**inky wrote:**

> For mounted units about half the time, and for drakes almost every single time. If it does happen it should be pretty noticeable.

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-08-05**

**vultraz wrote:**

> This probably has something to do with the framebuffer not being swapped enough.

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-08-06**

**wedge009 wrote:**

> What does this mean? Is it dependent on CPU performance? I can't say I've seen this happening before - at all.

**Comment by Coffee--** [comment]
As someone who was previously pretty familiar with the animation code, I've done some testing of this issue and found that it is a problem with a lua event implementation and some code being removed. Obscure, but this is how it seems to work:

In game_events/pump.cpp
wml_event_pump::operator()()
the code for the lua events that increments the impl_->internal_wml_tracking number (twice between each movement step it seems) changes the wml_tracking variable in actions/move.cpp unit_mover::do_move. When the wml_tracking doesn't match the do_move_track_ local variable because of the above lua events code the animator.proceed_to "animate" variable is set to true and the animations are reloaded at every step of the movement.

This results in choppy animations for movement and also maybe a performance hit as extra "fake units" are being created at every step instead of just for each movement path.

I've assigned this to you gfgtdf as it looks like you were the last to change this code.

The other part with artifacts I believe could be a couple of things, such as deleted code from unit_mover::wait_for_anims() function and maybe a performance hit from refreshing the screen instead of a specific location in the actions/move.cpp unit_mover::do_move the line:
move_it_->anim_comp().invalidate(disp);
which was in wesnoth 1.12 just invalidating the hex where the unit.

**Comment by Coffee--** [comment]
After some testing I've returned the mentioned 1.12 code in unit_mover::wait_for_anims() as this removes artifacts on screen for large units on movement (such as drakes).

The animations are still not smooth though and will need more work.

**Comment by gfgtdf** [comment]
hmm ok after my discussion with coffe yesterday, it seems like:

1) Whenever a wml event handler gets executed during moving (in particular enter_hex/exit_hex events) the animation dummy unit is rebuild to reflect possible changes to the unit done by wml code. This also 'resets' the animation state.

2) in 1.13 there is _always_ a a exit_hex /enter_hex event handler since the lua event handler (wesnoth.game_events.on_event) cannot distunguish beteen different event types.

In conclusion in 1.13.8 the animation state is reset after each hex resulting in graphical annoyances.

The best fix is probably to do a more careful tracking of when the unit really has changed and only to rebiuld the animation in that case.

**Comment by gfgtdf** [comment]
So i tried to make a list of whihc actions can change a unit appearance.

1) Obviously recreating the unit `[unstore_unit]`, `wesnoth.put_unit` etc can change the unit apperachance and required a rebuild of the unit animation.
2) `wesnoth.add_modification` , `wesnoth.remove_modifications`, `wesnoth.advance_unit` and `wesnoth.transform_unit`
3) I am unusure about whether changing unit stats (hp, xp etc) directly requires rebuiling

**Comment by CelticMinstrel** [comment]
I don't think changing stats would require rebuilding? They pretty much have no influence on the unit's appearance, right?

**Comment by gfgtdf** [comment]
hmm xe there is a bar that shpows these stats but it seems like it's not shown during movement.

**Comment by Coffee--** [comment]
I fixed all the non LUA engine stuff in commit ecb51029d75153eab311513e91c95246c59c9787

The unit still doesn't animate properly on move because of the LUA issue. Happy to leave that with you gfgtdf.

**Comment by stevecotton** [comment]
> I don't think changing stats would require rebuilding? They pretty much have no influence on the unit's appearance, right?

Generally, but not reliably; with "not reliably" meaning issue #4978. AnimationWML can use anything available to an SUF, but the `appearance_changed()` function added here is only checked when the unit is moving.
